File: mc/tools/check_evidence_manifests.py
#!/usr/bin/env python3
"""Validate tracked evidence BLAKE3 manifests and stale receipt markers."""
from __future__ import annotations

import argparse
import re
import shutil
import subprocess
import sys
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable

ROOT = Path(__file__).resolve().parents[1]
EVIDENCE_DIR = ROOT / "docs" / "evidence"
BLAKE3_HEX_LENGTH = 64
MANIFEST_GLOB = "*.b3"
RECEIPT_GLOB = "*.receipt.json"
MANIFEST_SEPARATOR = "  "
STALE_RECEIPT_MARKERS = {
    "equipment_packet_observed": "use entity_equipment_update",
}
BLAKE3_RE = re.compile(rf"^[0-9a-f]{{{BLAKE3_HEX_LENGTH}}}$")


@dataclass(frozen=True)
class ManifestEntry:
    manifest: Path
    digest: str
    relative_path: Path


@dataclass(frozen=True)
class CheckResult:
    manifests: int
    entries: int
    receipts_scanned: int


def repo_relative(path: Path, root: Path = ROOT) -> Path:
    return path.resolve().relative_to(root.resolve())


def parse_manifest(manifest: Path, text: str) -> tuple[list[ManifestEntry], list[str]]:
    entries: list[ManifestEntry] = []
    errors: list[str] = []
    for line_number, raw_line in enumerate(text.splitlines(), start=1):
        line = raw_line.strip()
        if not line:
            continue
        if MANIFEST_SEPARATOR not in line:
            errors.append(f"{manifest}:{line_number}: expected '<b3>  <path>'")
            continue
        digest, path_text = line.split(MANIFEST_SEPARATOR, maxsplit=1)
        if not BLAKE3_RE.fullmatch(digest):
            errors.append(f"{manifest}:{line_number}: invalid BLAKE3 digest {digest!r}")
            continue
        relative_path = Path(path_text.strip())
        if relative_path.is_absolute() or ".." in relative_path.parts:
            errors.append(f"{manifest}:{line_number}: path must be repo-relative: {path_text!r}")
            continue
        entries.append(ManifestEntry(manifest=manifest, digest=digest, relative_path=relative_path))
    if not entries and not errors:
        errors.append(f"{manifest}: manifest has no entries")
    return entries, errors


def discover_manifests(evidence_dir: Path = EVIDENCE_DIR) -> list[Path]:
    return sorted(evidence_dir.glob(MANIFEST_GLOB))


def discover_receipts(evidence_dir: Path = EVIDENCE_DIR) -> list[Path]:
    return sorted(evidence_dir.glob(RECEIPT_GLOB))


def validate_manifest_entries(entries: Iterable[ManifestEntry], root: Path = ROOT) -> list[str]:
    errors: list[str] = []
    for entry in entries:
        target = root / entry.relative_path
        if not target.exists():
            errors.append(f"{entry.manifest}: referenced file missing: {entry.relative_path}")
        elif not target.is_file():
            errors.append(f"{entry.manifest}: referenced path is not a file: {entry.relative_path}")
    return errors


def run_b3sum_check(manifests: Iterable[Path], root: Path = ROOT) -> list[str]:
    errors: list[str] = []
    if shutil.which("b3sum") is None:
        return ["b3sum not found on PATH; run under nix develop or Nix check"]
    for manifest in manifests:
        completed = subprocess.run(
            ["b3sum", "--check", str(repo_relative(manifest, root))],
            cwd=root,
            text=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            check=False,
        )
        if completed.returncode != 0:
            output = completed.stdout.strip() or "<no output>"
            errors.append(f"{manifest}: b3sum --check failed:\n{output}")
    return errors


def referenced_paths(entries: Iterable[ManifestEntry], root: Path = ROOT) -> set[Path]:
    return {(root / entry.relative_path).resolve() for entry in entries}


def receipt_paths(
    entries: Iterable[ManifestEntry],
    root: Path = ROOT,
    evidence_dir: Path = EVIDENCE_DIR,
) -> list[Path]:
    paths = referenced_paths(entries, root)
    paths.update(path.resolve() for path in discover_receipts(evidence_dir))
    return sorted(path for path in paths if path.name.endswith(".receipt.json"))


def stale_marker_errors(receipts: Iterable[Path]) -> list[str]:
    errors: list[str] = []
    for receipt in receipts:
        if not receipt.exists():
            continue
        text = receipt.read_text()
        for marker, replacement in sorted(STALE_RECEIPT_MARKERS.items()):
            if marker in text:
                errors.append(f"{receipt}: stale marker {marker!r}; {replacement}")
    return errors


def check_evidence(root: Path = ROOT, evidence_dir: Path = EVIDENCE_DIR) -> tuple[CheckResult, list[str]]:
    manifests = discover_manifests(evidence_dir)
    all_entries: list[ManifestEntry] = []
    errors: list[str] = []
    for manifest in manifests:
        entries, manifest_errors = parse_manifest(manifest, manifest.read_text())
        all_entries.extend(entries)
        errors.extend(manifest_errors)
    errors.extend(validate_manifest_entries(all_entries, root))
    errors.extend(run_b3sum_check(manifests, root))
    receipts = receipt_paths(all_entries, root, evidence_dir)
    errors.extend(stale_marker_errors(receipts))
    result = CheckResult(
        manifests=len(manifests),
        entries=len(all_entries),
        receipts_scanned=len(receipts),
    )
    return result, errors


def assert_self_tests() -> None:
    with tempfile.TemporaryDirectory(prefix="evidence-manifest-self-test-") as tmp:
        root = Path(tmp)
        evidence = root / "docs" / "evidence"
        evidence.mkdir(parents=True)
        receipt = evidence / "good.receipt.json"
        receipt.write_text('{"required_milestones":["entity_equipment_update"]}\n')
        manifest = evidence / "good.b3"
        digest = subprocess.check_output(["b3sum", str(receipt)], text=True).split()[0]
        manifest.write_text(f"{digest}{MANIFEST_SEPARATOR}docs/evidence/good.receipt.json\n")

        result, errors = check_evidence(root, evidence)
        assert not errors, errors
        assert result.manifests == 1, result
        assert result.entries == 1, result
        assert result.receipts_scanned == 1, result

        stale = evidence / "stale.receipt.json"
        stale.write_text('{"missing_milestones":["equipment_packet_observed"]}\n')
        _, errors = check_evidence(root, evidence)
        assert any("equipment_packet_observed" in error for error in errors), errors
        stale.unlink()

        run_log = evidence / "run.log"
        run_log.write_text("ok\n")
        run_log_digest = subprocess.check_output(["b3sum", str(run_log)], text=True).split()[0]
        log_manifest = evidence / "run-log.b3"
        log_manifest.write_text(f"{run_log_digest}{MANIFEST_SEPARATOR}docs/evidence/run.log\n")
        result, errors = check_evidence(root, evidence)
        assert not errors, errors
        assert result.entries == 2, result

        run_log.unlink()
        _, errors = check_evidence(root, evidence)
        assert any("referenced file missing" in error and "run.log" in error for error in errors), errors
        log_manifest.unlink()

        missing_manifest = evidence / "missing.b3"
        missing_manifest.write_text(
            f"{'0' * BLAKE3_HEX_LENGTH}{MANIFEST_SEPARATOR}docs/evidence/missing.receipt.json\n"
        )
        _, errors = check_evidence(root, evidence)
        assert any("referenced file missing" in error for error in errors), errors

        receipt.write_text('{"required_milestones":["entity_equipment_update"],"changed":true}\n')
        _, errors = check_evidence(root, evidence)
        assert any("b3sum --check failed" in error for error in errors), errors


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--self-test", action="store_true", help="run positive and negative fixtures")
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    if args.self_test:
        assert_self_tests()
        print("evidence manifest self-test ok")
        return 0
    result, errors = check_evidence()
    if errors:
        for error in errors:
            print(error, file=sys.stderr)
        return 1
    print(
        "evidence manifests ok: "
        f"{result.manifests} manifests, {result.entries} entries, "
        f"{result.receipts_scanned} receipts scanned"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: mc/tools/check_evidence_manifests.rs
use std::collections::BTreeSet;
use std::env;
use std::ffi::OsStr;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::{Command, ExitCode};

const SELF_TEST_FLAG: &str = "--self-test";
const EVIDENCE_DIR: &str = "docs/evidence";
const MANIFEST_EXTENSION: &str = "b3";
const RECEIPT_SUFFIX: &str = ".receipt.json";
const MANIFEST_SEPARATOR: &str = "  ";
const BLAKE3_HEX_LENGTH: usize = 64;
const SUCCESS: ExitCode = ExitCode::SUCCESS;
const FAILURE: ExitCode = ExitCode::FAILURE;
const STALE_RECEIPT_MARKER: &str = "equipment_packet_observed";
const STALE_RECEIPT_REPLACEMENT: &str = "use entity_equipment_update";
const TEMP_DIR_PREFIX: &str = "evidence-manifest-self-test-";
const CHECKER_NAME: &str = "evidence manifest";

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
struct ManifestEntry {
    manifest: PathBuf,
    digest: String,
    relative_path: PathBuf,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct CheckSummary {
    manifests: usize,
    entries: usize,
    receipts_scanned: usize,
}

fn main() -> ExitCode {
    let args = env::args().collect::<Vec<_>>();
    if args.iter().any(|arg| arg == SELF_TEST_FLAG) {
        return match run_self_tests() {
            Ok(summary) => {
                println!("{CHECKER_NAME} self-test ok: {summary}");
                SUCCESS
            }
            Err(errors) => {
                print_errors(&errors);
                FAILURE
            }
        };
    }

    match check_evidence(Path::new("."), Path::new(EVIDENCE_DIR)) {
        Ok(summary) => {
            println!(
                "evidence manifests ok: {} manifests, {} entries, {} receipts scanned",
                summary.manifests, summary.entries, summary.receipts_scanned
            );
            SUCCESS
        }
        Err(errors) => {
            print_errors(&errors);
            FAILURE
        }
    }
}

fn print_errors(errors: &[String]) {
    for error in errors {
        eprintln!("{CHECKER_NAME} check failed: {error}");
    }
}

fn check_evidence(root: &Path, evidence_dir: &Path) -> Result<CheckSummary, Vec<String>> {
    let manifests = discover_by_extension(evidence_dir, MANIFEST_EXTENSION);
    let mut entries = Vec::new();
    let mut errors = Vec::new();
    for manifest in &manifests {
        match fs::read_to_string(manifest) {
            Ok(text) => entries.extend(parse_manifest(manifest, &text, &mut errors)),
            Err(error) => errors.push(format!("{}: {error}", manifest.display())),
        }
    }
    errors.extend(validate_manifest_entries(root, &entries));
    errors.extend(run_b3sum_checks(root, &manifests));
    let receipts = receipt_paths(root, evidence_dir, &entries);
    errors.extend(stale_marker_errors(&receipts));

    let summary = CheckSummary {
        manifests: manifests.len(),
        entries: entries.len(),
        receipts_scanned: receipts.len(),
    };
    if errors.is_empty() {
        Ok(summary)
    } else {
        Err(errors)
    }
}

fn parse_manifest(manifest: &Path, text: &str, errors: &mut Vec<String>) -> Vec<ManifestEntry> {
    let mut entries = Vec::new();
    for (line_index, raw_line) in text.lines().enumerate() {
        let line_number = line_index + 1;
        let line = raw_line.trim();
        if line.is_empty() {
            continue;
        }
        let Some((digest, path_text)) = line.split_once(MANIFEST_SEPARATOR) else {
            errors.push(format!(
                "{}:{line_number}: expected '<b3>  <path>'",
                manifest.display()
            ));
            continue;
        };
        if !is_blake3_digest(digest) {
            errors.push(format!(
                "{}:{line_number}: invalid BLAKE3 digest {digest:?}",
                manifest.display()
            ));
            continue;
        }
        let relative_path = PathBuf::from(path_text.trim());
        if relative_path.is_absolute()
            || relative_path
                .components()
                .any(|part| part.as_os_str() == OsStr::new(".."))
        {
            errors.push(format!(
                "{}:{line_number}: path must be repo-relative: {path_text:?}",
                manifest.display()
            ));
            continue;
        }
        entries.push(ManifestEntry {
            manifest: manifest.to_path_buf(),
            digest: digest.to_string(),
            relative_path,
        });
    }
    if entries.is_empty() && errors.is_empty() {
        errors.push(format!("{}: manifest has no entries", manifest.display()));
    }
    entries
}

fn is_blake3_digest(value: &str) -> bool {
    value.len() == BLAKE3_HEX_LENGTH && value.bytes().all(|byte| byte.is_ascii_hexdigit())
}

fn discover_by_extension(dir: &Path, extension: &str) -> Vec<PathBuf> {
    let mut paths = Vec::new();
    if let Ok(entries) = fs::read_dir(dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_file() && path.extension() == Some(OsStr::new(extension)) {
                paths.push(path);
            }
        }
    }
    paths.sort();
    paths
}

fn discover_receipts(evidence_dir: &Path) -> Vec<PathBuf> {
    let mut paths = Vec::new();
    if let Ok(entries) = fs::read_dir(evidence_dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_file() && path.to_string_lossy().ends_with(RECEIPT_SUFFIX) {
                paths.push(path);
            }
        }
    }
    paths.sort();
    paths
}

fn validate_manifest_entries(root: &Path, entries: &[ManifestEntry]) -> Vec<String> {
    let mut errors = Vec::new();
    for entry in entries {
        let target = root.join(&entry.relative_path);
        if !target.exists() {
            errors.push(format!(
                "{}: referenced file missing: {}",
                entry.manifest.display(),
                entry.relative_path.display()
            ));
        } else if !target.is_file() {
            errors.push(format!(
                "{}: referenced path is not a file: {}",
                entry.manifest.display(),
                entry.relative_path.display()
            ));
        }
    }
    errors
}

fn run_b3sum_checks(root: &Path, manifests: &[PathBuf]) -> Vec<String> {
    let mut errors = Vec::new();
    for manifest in manifests {
        let relative_manifest = match manifest.strip_prefix(root) {
            Ok(path) => path,
            Err(_) => manifest.as_path(),
        };
        let output = Command::new("b3sum")
            .arg("--check")
            .arg(relative_manifest)
            .current_dir(root)
            .output();
        match output {
            Ok(completed) if completed.status.success() => {}
            Ok(completed) => {
                let mut text = String::from_utf8_lossy(&completed.stdout).to_string();
                text.push_str(&String::from_utf8_lossy(&completed.stderr));
                let trimmed = text.trim();
                let output_text = if trimmed.is_empty() {
                    "<no output>"
                } else {
                    trimmed
                };
                errors.push(format!(
                    "{}: b3sum --check failed:\n{output_text}",
                    manifest.display()
                ));
            }
            Err(error) => errors.push(format!("b3sum failed for {}: {error}", manifest.display())),
        }
    }
    errors
}

fn receipt_paths(root: &Path, evidence_dir: &Path, entries: &[ManifestEntry]) -> Vec<PathBuf> {
    let mut paths = BTreeSet::new();
    for path in discover_receipts(evidence_dir) {
        paths.insert(canonical_path(&path));
    }
    for entry in entries {
        if entry
            .relative_path
            .to_string_lossy()
            .ends_with(RECEIPT_SUFFIX)
        {
            paths.insert(canonical_path(&root.join(&entry.relative_path)));
        }
    }
    paths.into_iter().collect()
}

fn canonical_path(path: &Path) -> PathBuf {
    path.canonicalize().unwrap_or_else(|_| path.to_path_buf())
}

fn stale_marker_errors(receipts: &[PathBuf]) -> Vec<String> {
    let mut errors = Vec::new();
    for receipt in receipts {
        if !receipt.exists() {
            continue;
        }
        match fs::read_to_string(receipt) {
            Ok(text) if text.contains(STALE_RECEIPT_MARKER) => errors.push(format!(
                "{}: stale marker {STALE_RECEIPT_MARKER:?}; {STALE_RECEIPT_REPLACEMENT}",
                receipt.display()
            )),
            Ok(_) => {}
            Err(error) => errors.push(format!("{}: {error}", receipt.display())),
        }
    }
    errors
}

fn run_self_tests() -> Result<String, Vec<String>> {
    let root = make_temp_root()?;
    let evidence_dir = root.join(EVIDENCE_DIR);
    fs::create_dir_all(&evidence_dir)
        .map_err(|error| vec![format!("{}: {error}", evidence_dir.display())])?;

    let receipt = evidence_dir.join("good.receipt.json");
    write_file(
        &receipt,
        "{\"required_milestones\":[\"entity_equipment_update\"]}\n",
    )?;
    let manifest = evidence_dir.join("good.b3");
    let receipt_digest = b3sum_digest(&receipt)?;
    write_file(
        &manifest,
        &format!("{receipt_digest}{MANIFEST_SEPARATOR}docs/evidence/good.receipt.json\n"),
    )?;

    let summary = check_evidence(&root, &evidence_dir)?;
    assert_summary(
        summary,
        CheckSummary {
            manifests: 1,
            entries: 1,
            receipts_scanned: 1,
        },
    )?;

    let stale = evidence_dir.join("stale.receipt.json");
    write_file(
        &stale,
        "{\"missing_milestones\":[\"equipment_packet_observed\"]}\n",
    )?;
    assert_contains(&check_evidence(&root, &evidence_dir), STALE_RECEIPT_MARKER)?;
    fs::remove_file(&stale).map_err(|error| vec![format!("{}: {error}", stale.display())])?;

    let run_log = evidence_dir.join("run.log");
    write_file(&run_log, "ok\n")?;
    let run_log_digest = b3sum_digest(&run_log)?;
    let log_manifest = evidence_dir.join("run-log.b3");
    write_file(
        &log_manifest,
        &format!("{run_log_digest}{MANIFEST_SEPARATOR}docs/evidence/run.log\n"),
    )?;
    let summary = check_evidence(&root, &evidence_dir)?;
    assert_summary(
        summary,
        CheckSummary {
            manifests: 2,
            entries: 2,
            receipts_scanned: 1,
        },
    )?;

    fs::remove_file(&run_log).map_err(|error| vec![format!("{}: {error}", run_log.display())])?;
    assert_contains(
        &check_evidence(&root, &evidence_dir),
        "referenced file missing",
    )?;
    fs::remove_file(&log_manifest)
        .map_err(|error| vec![format!("{}: {error}", log_manifest.display())])?;

    let missing_manifest = evidence_dir.join("missing.b3");
    write_file(
        &missing_manifest,
        &format!(
            "{}{MANIFEST_SEPARATOR}docs/evidence/missing.receipt.json\n",
            "0".repeat(BLAKE3_HEX_LENGTH)
        ),
    )?;
    assert_contains(
        &check_evidence(&root, &evidence_dir),
        "referenced file missing",
    )?;

    write_file(
        &receipt,
        "{\"required_milestones\":[\"entity_equipment_update\"],\"changed\":true}\n",
    )?;
    assert_contains(
        &check_evidence(&root, &evidence_dir),
        "b3sum --check failed",
    )?;

    let _ = fs::remove_dir_all(&root);
    Ok("positive and negative fixtures exercised".to_string())
}

fn make_temp_root() -> Result<PathBuf, Vec<String>> {
    let mut root = env::temp_dir();
    root.push(format!("{TEMP_DIR_PREFIX}{}", std::process::id()));
    if root.exists() {
        fs::remove_dir_all(&root).map_err(|error| vec![format!("{}: {error}", root.display())])?;
    }
    fs::create_dir_all(&root).map_err(|error| vec![format!("{}: {error}", root.display())])?;
    Ok(root)
}

fn write_file(path: &Path, text: &str) -> Result<(), Vec<String>> {
    fs::write(path, text).map_err(|error| vec![format!("{}: {error}", path.display())])
}

fn b3sum_digest(path: &Path) -> Result<String, Vec<String>> {
    let output = Command::new("b3sum")
        .arg(path)
        .output()
        .map_err(|error| vec![format!("b3sum {}: {error}", path.display())])?;
    if !output.status.success() {
        return Err(vec![format!(
            "b3sum {} exited with {}",
            path.display(),
            output.status
        )]);
    }
    let stdout = String::from_utf8_lossy(&output.stdout);
    stdout
        .split_whitespace()
        .next()
        .map(ToOwned::to_owned)
        .ok_or_else(|| vec![format!("b3sum {} produced no digest", path.display())])
}

fn assert_summary(actual: CheckSummary, expected: CheckSummary) -> Result<(), Vec<String>> {
    if actual == expected {
        Ok(())
    } else {
        Err(vec![format!(
            "summary mismatch: got {actual:?}, expected {expected:?}"
        )])
    }
}

fn assert_contains(
    result: &Result<CheckSummary, Vec<String>>,
    needle: &str,
) -> Result<(), Vec<String>> {
    match result {
        Ok(summary) => Err(vec![format!(
            "expected diagnostic {needle:?}, got success {summary:?}"
        )]),
        Err(errors) if errors.iter().any(|error| error.contains(needle)) => Ok(()),
        Err(errors) => Err(vec![format!(
            "missing expected diagnostic {needle:?}: {errors:?}"
        )]),
    }
}
